src/aliceVision/sfm/pipeline/sequential/ReconstructionEngine_sequentialSfM.cpp

```diff
@@ -1415,59 +1430,281 @@ bool ReconstructionEngine_sequentialSfM::Resection(const std::size_t viewIndex)
+{
+  ALICEVISION_LOG_DEBUG("Triangulating (mode: multi-view LO-RANSAC)... ");
+  
+  // -- Identify the track to triangulate :
+  // This map contains all the tracks that will be triangulated (for the first time, or not)
+  // These tracks are seen by at least one new reconstructed view.
+  std::map<IndexT, std::set<IndexT>> mapTracksToTriangulate; // <trackId, observations> 
+  std::vector<IndexT> setTracksId; // <trackId>
+  {
+#pragma omp parallel for 
+    for (ptrdiff_t i = 0; i < static_cast<ptrdiff_t>(previousReconstructedViews.size()); ++i)
+    {
+      std::set<IndexT>::const_iterator iter = previousReconstructedViews.begin();
+      std::advance(iter, i);
+      const IndexT indexPrev = *iter;
+      
+      for(IndexT indexNew: newReconstructedViews)
+      {
+        if(indexPrev == indexNew)
+          continue;
+        
+        const std::size_t I = std::min((IndexT)indexNew, indexPrev);
+        const std::size_t J = std::max((IndexT)indexNew, indexPrev);
+        
+        // Find track correspondences between I and J
+        const std::set<std::size_t> set_viewIndex = { I, J };
+        track::TracksMap map_tracksCommonIJ;
+        track::TracksUtilsMap::GetTracksInImagesFast(set_viewIndex, _map_tracks, _map_tracksPerView, map_tracksCommonIJ);
+        
+        // Collect tracksIds
+        std::set<std::size_t> commonTracksId;
+        std::transform(map_tracksCommonIJ.begin(), map_tracksCommonIJ.end(),
+                       std::inserter(commonTracksId, commonTracksId.begin()),
+                       stl::RetrieveKey());
+        
+        for (std::size_t trackId : commonTracksId)
+        {
+#pragma omp critical        
+          {
+            mapTracksToTriangulate[trackId].insert(I);      
+            mapTracksToTriangulate[trackId].insert(J);      
+          }
+        } // tracks to reconstruct/update
+      } // new reconstructed view
+    } // all recontrusted views
+#pragma omp critical        
+    {
+      std::transform(mapTracksToTriangulate.begin(), mapTracksToTriangulate.end(),
+                     std::inserter(setTracksId, setTracksId.begin()),
+                     stl::RetrieveKey());
+    }
+  }
+  
+#pragma omp parallel for 
+  for (int i = 0; i < setTracksId.size(); i++) // each track (already reconstructed or not)
+  {
+    const IndexT trackId = setTracksId.at(i);
+    const track::Track & track = _map_tracks.at(trackId);
+    std::set<IndexT> & observations = mapTracksToTriangulate.at(trackId); // all the posed views possessing the track
+    
+    // The track needs to be seen by a min. number of views to be triangulated
+    if (observations.size() < _minNbObservationsForTriangulation)
+      continue;
+    
+    Vec3 X_euclidean = Vec3::Zero();
+    std::set<IndexT> inliers;
+    
+#pragma omp critical
+    {
+      if (scene.structure.find(trackId) != scene.structure.end()) 
+      {
+        scene.structure.erase(trackId);
```

Comment:
Here we delete it at the beginning and if it is still valid at the end (which is the prevalent case) we add it again.
We can simply update the value or remove it explicitly if we consider it is as an outlier. So we always have 1 critical section instead of 2.